### ISSUE TITLE
Allow more APIs for CodeQuestion

### ISSUE THREAD

**Issue by dzhuang** [inducer/relate#688]
(Hi, long time no see.)

I wish to added more APIs for CodeQuestions, so that some simple tests could be done which is currently not possible for relate AFAIK. For example, testing if a specifc function was called with specific arguments.

Another use case is to test scripts (not a function) submit by a user. For example, I need to test if `input` is used in a script. The following was the `user_code`:

```python
name = input("please input your name")
print(”Hello %s!" % name)
```

My workround is to turn `user_code` into the following before sent to [request_run](https://github.com/inducer/relate/blob/68ecca7ea440ee4eaf70312168a54e5ecf73aa37/course/page/code.py#L602):
```python
def __func():
      name = input("please input your name")
      print(”Hello %s!" % name)
```
and then I can use unit test to check whether it really works, using something like [this](https://forum.learncodethehardway.com/t/testing-input-and-print/1757/10) using `mock`. That It also require that the `__func` namespace be added to the `run_req` as a runpy context so that `__func' can be called in `test_code`. 

To summarize, I need to refactor the [`grade`](https://github.com/inducer/relate/blob/68ecca7ea440ee4eaf70312168a54e5ecf73aa37/course/page/code.py#L595) method so that more possibility of subclass of `CodeQuestion` or `PythonCodeQuestion` is possible.

May I write a PR to do that? Thanks.

**Comment by inducer** [comment]
Good to see you again @dzhuang!

If I understand correctly, what you're requesting can (mostly) already be done, as Relate provides the source code from the participant to the grading code (as a string) in `user_code`. One particular tweak that might be needed is to introduce a knob that prevents Relate from trying to run that code. (So that it can be run only under the control of the grading code.) I'm open to consider that change. This should enforce that `names_{for,from}_user` is empty.

**Comment by dzhuang** [comment]
Thank you for your response and sorry for my late response (17 days passed :<). I'll try to submit a PR when I'm ready.
